lib/reversetunnel/localsite.go

```diff
@@ -544,54 +573,94 @@ func (s *localSite) handleHeartbeat(rconn *remoteConn, ch ssh.Channel, reqC <-ch
-			} else {
-				s.log.WithFields(log.Fields{"nodeID": rconn.nodeID}).Debugf("Ping <- %v", rconn.conn.RemoteAddr())
+				log = logger.WithField("latency", roundtrip)
 			}
-			tm := time.Now().UTC()
-			rconn.setLastHeartbeat(tm)
+			log.Debugf("Ping <- %v", rconn.conn.RemoteAddr())
+
+			rconn.setLastHeartbeat(s.clock.Now().UTC())
+			rconn.markValid()
 		// Note that time.After is re-created everytime a request is processed.
-		case <-time.After(s.offlineThreshold):
+		case t := <-s.clock.After(s.offlineThreshold):
 			rconn.markInvalid(trace.ConnectionProblem(nil, "no heartbeats for %v", s.offlineThreshold))
+
+			// terminate and remove the connection after missing more than missedHeartBeatThreshold heartbeats if
+			// the connection isn't still servicing any sessions
+			hb := rconn.getLastHeartbeat()
+			if t.After(hb.Add(s.offlineThreshold * missedHeartBeatThreshold)) {
+				count := rconn.activeSessions()
+				if count == 0 {
+					logger.Errorf("Closing unhealthy and idle connection. Heartbeat last received at %s", hb)
+					return
+				}
+
+				logger.Warnf("Deferring closure of unhealthy connection due to %d active connections", count)
+			}
 		}
 	}
 }
 
-func (s *localSite) getRemoteConn(dreq *sshutils.DialReq) (*remoteConn, error) {
+func (s *localSite) removeRemoteConn(rconn *remoteConn) {
 	s.remoteConnsMtx.Lock()
 	defer s.remoteConnsMtx.Unlock()
 
-	// Loop over all connections and remove and invalid connections from the
-	// connection map.
-	for key, conns := range s.remoteConns {
-		validConns := conns[:0]
-		for _, conn := range conns {
-			if !conn.isInvalid() {
-				validConns = append(validConns, conn)
+	key := connKey{
+		uuid:     rconn.nodeID,
+		connType: types.TunnelType(rconn.tunnelType),
+	}
+
+	conns := s.remoteConns[key]
+	for i, conn := range conns {
+		if conn == rconn {
+			s.remoteConns[key] = append(conns[:i], conns[i+1:]...)
+			if len(s.remoteConns[key]) == 0 {
+				delete(s.remoteConns, key)
 			}
-		}
-		if len(validConns) == 0 {
-			delete(s.remoteConns, key)
-		} else {
-			s.remoteConns[key] = validConns
+			return
 		}
 	}
+}
+
+func (s *localSite) getRemoteConn(dreq *sshutils.DialReq) (*remoteConn, error) {
+	s.remoteConnsMtx.Lock()
+	defer s.remoteConnsMtx.Unlock()
 
 	key := connKey{
 		uuid:     dreq.ServerID,
 		connType: dreq.ConnType,
 	}
-	if len(s.remoteConns[key]) == 0 {
+
+	conns := s.remoteConns[key]
+	if len(conns) == 0 {
 		return nil, trace.NotFound("no %v reverse tunnel for %v found", dreq.ConnType, dreq.ServerID)
 	}
 
-	conns := s.remoteConns[key]
+	// Check the remoteConns from newest to oldest for one
+	// that has heartbeated and is valid. If none are valid, try
+	// the newest ready but invalid connection.
+	var newestInvalidConn *remoteConn
 	for i := len(conns) - 1; i >= 0; i-- {
-		if conns[i].isReady() {
+		switch {
+		case !conns[i].isReady(): // skip remoteConn that haven't heartbeated yet
+			continue
```

Comment:
Did some poking around... doesn't look like waiting for the first *heartbeat* is necessary, but it does seem like it is necessary to wait for the heartbeat *channel* to be created, just to make sure that the agent has set up all its request/channel handlers.  Probably best to leave the logic here as-is.

In theory, I think we could eek out a bit of an improvement in setup time by just getting rid of the first hb sent by the agent, and treating the hb channel open *as* the first hb.  Out of scope for this PR I think, but something worth remembering if we ever decide tunnel setup time is a problem in the future.

Reply:
It wouldn't be much of an improvement, channel open and first heartbeat are one after another in the same goroutine with no extra sleeps.

I agree that the future protocol should tie the availability of the node side for new connections to the lifetime of a `reversetunnel@goteleport.com` channel rather than to some weird heartbeat system (for a couple of protocol versions, then we can decide if that should be a global request that irreversibly marks that ssh connection as available for reverse tunneling, or if we just leave it as is and figure out some TLS based way of doing it).

Reply:
On said hypothetical new reverse tunnel protocol we should just remove the entire heartbeat mechanism. The only value that it gives us is that an agent that's stopping is going to stop heartbeating and it's going to reject new connection attempts, so _maybe_ there's a point to treating a connection that's not heartbeating as a "lesser" one; however, the only true indicators that a connection is not actually usable for new connections anymore is a ssh error (potentially caused by closing the underlying connection due to idle timeouts?) or a reject on a `teleport-transport` channel open.